!!! TASK: Remove leftovers from magic underscore access for internal node properties

mhsdesign opened the pull request:
We have the `_*` underscore access syntax to access _actual_ php object fields of the Node. Due to the ESCR rewrite the Node object looks completely different and has way less getters / fields. See https://github.com/neos/neos-development-collection/issues/4208#issuecomment-2372012098 for a comparison.

The only overlap of fields to the old node are those few:

- `_name`
- `_nodeTypeName` (unknown as `.nodeType` was always used) 
- `_properties` (useless as `q(node).property(...)` is used)

Keeping the underscore access would expose a lot of new node fields like `_contentRepositoryId` which has no gain over to `node.contentRepositoryId`.
Adding a full runtime migration layer for fields of 8.3 like `_identifier`, `_depth`, `_nodeType` and `_hiddenAfterDateTime` was decided against as there is a rector migration in place.

So to avoid confusion about which magic underscore properties still exist in 9.0 - as rule of thumb - None, and the syntax will be removed from fusion. One can use an underscore now in consumer land but it is never any more special than any other character.

The sort operation previously also handled `_creationDateTime`, `_lastModificationDateTime` and `_lastPublicationDateTime`. For Neos 9.0 the new `sortByTimestamp(created|lastModified|originalCreated|originalLastModified)` flowquery can be used.

Related: #4921
Resolves: #4208

mhsdesign committed:
TASK: Provide helper to find out if node is disabled in fusion

As we don't support property('_hidden') there would be no way to figure out if a node is hidden.

mhsdesign committed:
!!! TASK: Remove underscore node property access `_*` from property and filter FlowQuery

Fixes https://github.com/neos/neos-development-collection/issues/4208#issuecomment-1573345069

mhsdesign committed:
TASK: Remove legacy b/c for `property('_path')`

This migration is handled via rector:
https://github.com/neos/rector/blob/4843c45f5bd901bd63aad0d1d0367df394e2cd1e/config/set/contentrepository-90.php#L174C25-L174C45

mhsdesign committed:
TASK: Remove legacy b/c for `property('_identifier')`

This migration is handled via rector:
https://github.com/neos/rector/blob/4843c45f5bd901bd63aad0d1d0367df394e2cd1e/config/set/contentrepository-90.php#L191C25-L191C51

mhsdesign committed:
TASK: Move orphaned `SortOperation` to other Node eel flow-queries

mhsdesign committed:
!!! TASK: Remove underscore node property access `sort('_*')` from sort FlowQuery

mhsdesign commented:
one todo is: how to sort by creation date in Neos 9?? Something like node.timestamps.originalCreated
https://github.com/neos/neos-development-collection/issues/4208#issuecomment-1986006454

in our weekly from the 26. we discussed that instead of adding a new flowquery or a magic parameter, we probably want to support the old legacy 8.3 syntax `_creationDateTime` or `_lastModificationDateTime` or even `_lastPublicationDateTime`.

The question is if these modifiers should only work for the `sort` operation or also for `property` and `filter` as well as part of a fizzle query `[_creationDateTime = 123]`. (Fizzle is almost impossible to archive with the optimisation https://github.com/neos/neos-development-collection/pull/4712 in place)

mhsdesign commented:
todos

- [x] as we relocate the sort operation to the other flow-queries and touch its logic i seem to have to carry out the burden to migrate the old functional test https://github.com/neos/neos-development-collection/blob/fe0063b7e4fbaea9b0a72abfd323354fabf42173/Neos.Neos/Tests/Functional/FlowQueryOperations/SortOperationTest.php to behat
- [x] rector migration for `node.hidden` and `_hidden` to `Neos.Node.isDisabled(node)` https://github.com/neos/rector/pull/83

kitsunet commented:
I do not understand why you did this?

mhsdesign commented:
That was what i have been discussing in https://github.com/neos/neos-development-collection/issues/4208 or in the meetings i thought but it seems we have talked past each other?

My thesis in short (now part of the main description):

We have the `_*` underscore access syntax to access _actual_ php object properties of the Node. Due to the ESCR rewrite the Node object looks completely different and has way less getters / properties. These are the new properties:

- _subgraphIdentity
- _nodeAggregateId
- _originDimensionSpacePoint
- _classification
- _nodeTypeName
- _nodeName
- _timestamps

While these are the ones known from 8.3

- _nodeType (until https://github.com/neos/neos-development-collection/issues/5019)
- _properties (useless)

The actual important and ones used in 8.3 like `_identifier` `_depth` `_parent` `_hiddenAfterDateTime` `_hiddenBeforeDateTime ` etc (see https://github.com/neos/neos-development-collection/issues/4208#issuecomment-1573345069) will not work and have to migrated with rector.

To avoid confusion about which magic underscore properties still exist in 9.0 a simple way is - it thought we agreed upon this in the meeting from the 19.01 [link](https://github.com/neos/neos-development-collection/issues/4208#issuecomment-1900183948) - to NOT have any underscore property logic and remove any magic from this syntax. Like one can use an underscore now in consumer land but it is never any more special than any other character.

Thats why this pr remove partially implemented legacy support for `_identifier` as well as `_depth` as we have rector migrations for this to properly use the respective eel helper or eel property access directly.

There is a little but as discussed in on the 26.04 we concluded to still allow sorting nodes by date via the `q().sort()` we will likely have to make one exception regarding underscore properties: The sort operation will translate `_lastPublicationDateTime` and respective things to the 9.0 equivalent logic internally instead of having to introduce another  sort operation or a special parameter or another special syntax like `.sort(['$$lastModified'])` or such.

kitsunet commented:
This is all clear, but

> Extracted everything i though this pr was originally meant to be into https://github.com/neos/neos-development-collection/pull/5015 and added the things i found missing.

why, why didn't you just add the things missing to my original PR?

mhsdesign commented:
Because i thought it would get too complex and the `_hiddenInIndex` thing is somehow related but a different task and should imo deserve its own pr. (Also i asked you in slack if im allowed to do this :D)

kitsunet commented:
> Also i asked you in slack if im allowed to do this :D

Too much going on LOL, seems still a bit weird but 🤷

mhsdesign committed:
TASK: Adjust docs of `PropertyOperation`

mhsdesign committed:
TASK: Add behat test for sort operation

mhsdesign committed:
TASK: Readd unit test for sort operation

was previously removed via https://github.com/neos/neos-development-collection/pull/5036

mhsdesign committed:
TASK: Reimplement _creationDateTime, _lastModificationDateTime and _lastPublicationDateTime support for `sort`

mhsdesign committed:
TASK: Simplify happy path for q().property

We should not need to load the node type for every fetch, only if we want to fetch references

mhsdesign commented:
The migration to use the herby introduced `Neos.Node.isDisabled(node)` helper can be reviewed here: https://github.com/neos/rector/pull/83

This pr is ready for final reviews.

If i get feedback regarding "9.0 proposed getter and documentation" from above i can also add that documentation to the node.

mhsdesign committed:
Merge remote-tracking branch 'origin/9.0' into task/4208-remove-magic-underscore-access-for-internal-node-properties

dlubitz commented on Neos.ContentRepository.NodeAccess/Classes/FlowQueryOperations/SortOperation.php:
```diff
@@ -113,15 +111,12 @@ public function evaluate(FlowQuery $flowQuery, array $arguments)
-            if ($propertyValue instanceof \DateTime) {
-                $propertyValue = $propertyValue->getTimestamp();
-            }
+            $propertyValue = match($sortProperty) {
+                '_creationDateTime' => $node->timestamps->created->getTimestamp(),
+                '_lastModificationDateTime' => $node->timestamps->lastModified?->getTimestamp(),
+                '_lastPublicationDateTime' => $node->timestamps->originalLastModified?->getTimestamp(),
+                default => $node->getProperty($sortProperty)
+            };
```
I don't like this. If we wouldn't have this syntax in Neos <9, we would never come up with somethink like that.

Why not dedicated operations? `sortByCreationDateTime`, `sortByLastModificationDateTime`? 
Or at least `sortByTimeStamps(created|lastModified|originalLastModified)`?

kitsunet replied:
I like this idea!

mhsdesign replied:
well i would not oppose to that if this is what we agree on thats fine. I just know that christian also agreed with me on doing it like this

> As discussed briefly on the 26.04 we concluded to still allow sorting nodes by date via the `q().sort()` we will likely have to make one exception regarding underscore properties: The sort operation will translate `_lastModificationDateTime` and respective things to the 9.0 equivalent logic internally instead of having to introduce another  sort operation or a special parameter or another special syntax like `.sort(['$$lastModified'])` or such.

so just let me know if we should go down dennys proposed route then ill implement it.

mhsdesign replied:
introduced `sortByTimestamp` fq

mhsdesign committed:
TASK: Remove obsolete ParentsOperationTest

mhsdesign committed:
FEATURE: Introduce sortByTimestamp fq

dlubitz commented on Neos.ContentRepository.NodeAccess/Classes/FlowQueryOperations/SortByTimeStampOperation.php:
```diff
@@ -0,0 +1,99 @@
+use Neos\Eel\FlowQuery\Operations\AbstractOperation;
+
+/**
+ * "sortByTimeStamp" operation working on ContentRepository nodes.
+ * Sorts nodes by specified timestamp.
+ */
+class SortByTimeStampOperation extends AbstractOperation
```
Should be "Time**s**tamp", right?

mhsdesign replied:
i guess idk can adjust this ...

mhsdesign replied:
adjusted

mhsdesign committed:
TASK: Correct casing of `SortByTimeStampOperation`

dlubitz committed:
TASK: Correct casing of SortByTimestampOperation

bwaidelich commented on Neos.ContentRepository.NodeAccess/Classes/FlowQueryOperations/SortByTimestampOperation.php:
```diff
@@ -0,0 +1,99 @@
+            $timeStamp = match($arguments[0] ?? null) {
+                'created' => $node->timestamps->created->getTimestamp(),
+                'lastModified' => $node->timestamps->lastModified?->getTimestamp(),
+                'originalLastModified' => $node->timestamps->originalLastModified?->getTimestamp(),
```
There's also `originalCreated` – did you leave that out by intention?

mhsdesign replied:
ups because of moving this change around so much^^  - adjusted

bwaidelich reviewed (commented):
Looking good! Just a remark regarding the new sort operation

mhsdesign committed:
TASK: Add `originalCreated` as option

bwaidelich commented on Neos.Neos/Tests/Behavior/Features/Fusion/FlowQuery.feature:
```diff
@@ -395,6 +395,32 @@ Feature: Tests for the "Neos.ContentRepository" Flow Query methods.
+      unsorted = ${q([a1a3, a1a4, a1a1, a1a2]).get()}
+      sortByTitleAsc = ${q([a1a3, a1a4, a1a1, a1a2]).sort("title", "ASC").get()}
+      sortByUriDesc = ${q([a1a3, a1a4, a1a1, a1a2]).sort("uriPathSegment", "DESC").get()}
+      # todo find out how to test time related logic
```
We already do that in the Timestamps tests: https://github.com/neos/neos-development-collection/blob/9.0/Neos.ContentRepository.BehavioralTests/Tests/Behavior/Features/NodeTraversal/Timestamps.feature#L6

mhsdesign replied:
noooooin 😂 i didnt want to know now i have to test it 😂

mhsdesign replied:
Nu aber.

mhsdesign committed:
Merge remote-tracking branch 'origin/9.0' into task/4208-remove-magic-underscore-access-for-internal-node-properties

mhsdesign committed:
TASK: Test `sortByTimestamp` by manipulating date time

mhsdesign committed:
Merge remote-tracking branch 'origin/9.0' into task/4208-remove-magic-underscore-access-for-internal-node-properties

bwaidelich reviewed (approved):
Lovely, thanks!
